timemachine/cpp/src/nonbonded_pairs.cu

```diff
@@ -0,0 +1,197 @@
+#include "gpu_utils.cuh"
+#include "k_nonbonded_pairs.cuh"
+#include "math_utils.cuh"
+#include "nonbonded_pairs.hpp"
+#include <stdexcept>
+#include <vector>
+
+namespace timemachine {
+
+template <typename RealType, bool Negated, bool Interpolated>
+NonbondedPairs<RealType, Negated, Interpolated>::NonbondedPairs(
+    const std::vector<int> &pair_idxs,          // [M, 2]
+    const std::vector<double> &scales,          // [M, 2]
+    const std::vector<int> &lambda_plane_idxs,  // [N]
+    const std::vector<int> &lambda_offset_idxs, // [N]
+    const double beta,
+    const double cutoff,
+    const std::string &kernel_src)
+    : N_(lambda_offset_idxs.size()), M_(pair_idxs.size() / 2), beta_(beta), cutoff_(cutoff),
+      compute_w_coords_instance_(kernel_cache_.program(kernel_src.c_str()).kernel("k_compute_w_coords").instantiate()),
+      compute_permute_interpolated_(
+          kernel_cache_.program(kernel_src.c_str()).kernel("k_permute_interpolated").instantiate()),
+      compute_add_du_dp_interpolated_(
+          kernel_cache_.program(kernel_src.c_str()).kernel("k_add_du_dp_interpolated").instantiate()) {
+
+    if (pair_idxs.size() % 2 != 0) {
+        throw std::runtime_error("pair_idxs.size() must be exactly 2*M");
+    }
+
+    for (int i = 0; i < M_; i++) {
+        auto src = pair_idxs[i * 2 + 0];
```

Comment:
@proteneer Do we use auto? I thought you had told me not to do that, but could be mistaken

Reply:
(This is not new; it was copied from existing code in `nonbonded.cu`.)